### PR TITLE
disallow-untyped-defs in `docker` and `stubs` directories

### LINKED ISSUE #12190: Run mypy over docker directory
There's a couple of scripts in there which it would be nice to typecheck.

- change mypy.ini's `files` list to include `docker`
- run `mypy` and see what errors crop up
- fix them until none remain

There's the `contrib` directory too, but we don't promise to support that diligently.

### PR DESCRIPTION
**Pull request by DMRobertson**
Recommend commit-by-commit review.

I wasn't able to configure mypy to run with disallow-untyped-defs on these directories, so it'll be unenforced for now. (See also #11302.)

Resolves #12190.

### CODE REVIEW

File: docker/configure_workers_and_start.py [hunk 1]
```diff
@@ -547,7 +549,7 @@ def generate_worker_log_config(
     return log_config_filepath
 
 
-def main(args, environ):
+def main(args: List[str], environ: MutableMapping[str, str]) -> None:
```
Comment (squahtx) [inline docker/configure_workers_and_start.py:552]: I can't figure out why `environ` needs to be mutable?

File: stubs/txredisapi.pyi [hunk 2]
```diff
@@ -34,11 +36,11 @@ class RedisProtocol(protocol.Protocol):
...
 class SubscriberProtocol(RedisProtocol):
-    def __init__(self, *args, **kwargs): ...
+    def __init__(self, *args: object, **kwargs: object): ...
     password: Optional[str]
-    def subscribe(self, channels: Union[str, List[str]]): ...
-    def connectionMade(self): ...
-    def connectionLost(self, reason): ...
+    def subscribe(self, channels: Union[str, List[str]]) -> "Deferred[None]": ...
+    def connectionMade(self) -> None: ...
+    def connectionLost(self, reason: Failure = ...) -> None: ...
```
Comment (squahtx) [inline stubs/txredisapi.pyi:43]: Where does the default value for `reason` come from?

File: docker/configure_workers_and_start.py (hunk 1 above)
Comment (DMRobertson) [reply docker/configure_workers_and_start.py:552]: We mutate the `environ`, see e.g.g +559; I was trying to be a light touch here and avoid logic/behaviour changes.

(I was surprised to learn that the `os.environ` mapping is mutable!)

Comment (squahtx) [reply docker/configure_workers_and_start.py:552]: Ah, I missed that!

File: docker/start.py [hunk 3]
```diff
@@ -83,7 +89,7 @@ def generate_config_from_template(config_dir, config_path, environ, ownership):
 
     # Convert SYNAPSE_NO_TLS to boolean if exists
     if "SYNAPSE_NO_TLS" in environ:
-        tlsanswerstring = str.lower(environ["SYNAPSE_NO_TLS"])
+        tlsanswerstring = str.lower(environ["SYNAPSE_NO_TLS"])  # type
```
Comment (squahtx) [inline docker/start.py:92]: What does "# type" do?

File: stubs/txredisapi.pyi (hunk 2 above)
Comment (DMRobertson) [reply stubs/txredisapi.pyi:43]: By default it's a failure which holds a `ConnectionDone` instance. If you feel like spelunking:

- https://twistedmatrix.com/documents/current/api/twisted.internet.protocol.Protocol.html#connectionLost
- https://github.com/twisted/twisted/blob/5ed041274f70285299bcb3dcf11da70604416f63/src/twisted/internet/protocol.py#L571
- https://github.com/twisted/twisted/blob/5ed041274f70285299bcb3dcf11da70604416f63/src/twisted/internet/protocol.py#L524
- https://github.com/twisted/twisted/blob/5ed041274f70285299bcb3dcf11da70604416f63/src/twisted/internet/error.py#L221

Comment (DMRobertson) [reply stubs/txredisapi.pyi:43]: Actually, that is for the abstract `Protocol`. Let me double check that txredisapi doesn't override that somehow...

File: docker/start.py (hunk 3 above)
Comment (DMRobertson) [reply docker/start.py:92]: Nothing. I think this was a `#type: ignore` I started adding but thought better of it. Will remove, thanks!

File: stubs/txredisapi.pyi (hunk 2 above)
Comment (DMRobertson) [reply stubs/txredisapi.pyi:43]: Oh. It does: https://github.com/IlyaSkriblovsky/txredisapi/blob/c89a49829d1c48baa10d03c41d39c261c02ad6e1/txredisapi.py#L335. Moveover it seems like lots of the implementations of `Protocol` in twisted itself don't have a default for `why`.

In that case, let me remove the default argument.

File: docker/start.py (hunk 3 above)
Comment (DMRobertson) [reply docker/start.py]: https://github.com/matrix-org/synapse/pull/12528/commits/283d9614018a154f0570a43d2b2dde4dfac52b4c

File: stubs/txredisapi.pyi (hunk 2 above)
Comment (DMRobertson) [reply stubs/txredisapi.pyi]: https://github.com/matrix-org/synapse/pull/12528/commits/b4198640f0fad24af70a43c982ecfe7ed6dad460

**Review by squahtx — APPROVED** [review]
lgtm!

**Pull request merged**